service-base/src/main/java/org/eclipse/hono/service/amqp/RequestResponseEndpoint.java

```diff
@@ -237,54 +296,182 @@ protected final void handleMessage(final ProtonConnection con, final ProtonRecei
+                    String.format("reply-to address must have the following format %s/<tenant>/<reply-address>",
+                            getName())));
+            sender.close();
+            return;
+        }
+
+        final String replyTo = replyToAddress.toString();
+
+        if (this.replyAddresses.contains(replyTo)) {
+            logger.debug("client [{}] wanted to subscribe to already subscribed reply-to address [{}]",
+                    sender.getName(), replyTo);
+            sender.setCondition(ProtonHelper.condition(AmqpError.ILLEGAL_STATE,
+                    String.format("reply-to address [%s] is already subscribed", replyTo)));
             sender.close();
+            return;
+        }
+
+        logger.debug("establishing response sender link with client [{}]", sender.getName());
+
+        final MessageConsumer<JsonObject> replyConsumer = vertx.eventBus().consumer(replyTo,
+                message -> {
+                    // TODO check for correct session here...?
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("forwarding reply to client [{}]: {}", sender.getName(),
+                                message.body().encodePrettily());
+                    }
+                    final EventBusMessage response = EventBusMessage.fromJson(message.body());
+                    filterResponse(Constants.getClientPrincipal(con), response)
+                            .recover(t -> {
+                                final int status = ServiceInvocationException.extractStatusCode(t);
+                                return Future.succeededFuture(response.getResponse(status));
+                            })
+                            .map(filteredResponse -> {
+                                final Message amqpReply = getAmqpReply(filteredResponse);
+                                sender.send(amqpReply);
+                                flowCreditToRequestor(replyTo);
+                                return null;
+                            })
+                            .recover(t -> {
+                                flowCreditToRequestor(replyTo);
+                                return Future.failedFuture(t);
+                            });
+                });
+
+        // register this consumer and replyTo address with this connection
+
+        registerConsumerForConnection(con, replyTo, replyConsumer);
+
+        sender.setQoS(ProtonQoS.AT_LEAST_ONCE);
+
+        HonoProtonHelper.setCloseHandler(sender, senderClosed -> {
+            logger.debug("client [{}] closed sender link, removing associated event bus consumer [{}]",
+                    sender.getName(), replyConsumer.address());
+
+            deallocateReceiverForReplyTo(replyTo);
+            unregisterConsumerForConnection(con, replyTo, replyConsumer);
+
+            if (senderClosed.succeeded()) {
```

Comment:
I don't see why we shouldn't close the sender if it contains an error message (which we don't seem to care about anyway).

Reply:
Well that comes from the original code. I kept it the same way.

Reply:
can you change it anyway?

Reply:
Done